### PR TITLE
[to Move to Pilot Repository] Discover pilot parameters at runtime.

### PR DESCRIPTION
**Pull request by CinziaLu**
Discover pilot parameters (Commands, extensions and version) at runtime.
Removed the -X and -E command line options.
The json file is not created if it already exists.

For issue https://github.com/DIRACGrid/DIRAC/issues/2724

### CODE REVIEW

File: WorkloadManagementSystem/PilotAgent/dirac-pilot.py [hunk 1]
```diff
@@ -40,12 +40,42 @@
   pilotParams.pilotRootPath = os.getcwd()
   pilotParams.pilotScript = os.path.realpath( sys.argv[0] )
   pilotParams.pilotScriptName = os.path.basename( pilotParams.pilotScript )
+  grid = pilotParams.site.split( '.' )[0]
   log.debug( 'PARAMETER [%s]' % ', '.join( map( str, pilotParams.optList ) ) )
+  try:
+    import json
+  except ImportError:
+    log.error( 'No json module available, could not get pilot commands at runtime. Using the default list.' )
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/dirac-pilot.py]: Fine, but then you should not keep going from here: json is requested few lines below.

File: WorkloadManagementSystem/PilotAgent/dirac-pilot.py [hunk 2]
```diff
@@ -40,12 +40,42 @@
...
+  result = retrieveUrlTimeout( pilotParams.pilotCFGFileLocation + '/' + pilotParams.pilotCFGFile,
+                               pilotParams.pilotCFGFile,
+                               log,
+                               timeout = 120 )
+
+  if result:
+    fp = open( pilotParams.pilotCFGFile + '-local', 'r' )
+    pilotCommandsFileContent = json.load( fp )
+    fp.close()
+    if pilotParams.setup in pilotCommandsFileContent.keys():
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/dirac-pilot.py]: I think .keys() is not needed. Nor here, nor in later lines

File: WorkloadManagementSystem/PilotAgent/dirac-pilot.py [hunk 3]
```diff
@@ -40,12 +40,42 @@
...
+      elif grid in pilotCommandsFileContent['Defaults']['Commands'].keys():
+        pilotParams.commands = [str( pv ) for pv in pilotCommandsFileContent['Defaults']['Commands'][grid]]
+      else:
+        pilotParams.commands = [str( pv ) for pv in pilotCommandsFileContent['Defaults']['Commands']['defaultList']]
+      if 'Extensions' in pilotCommandsFileContent[pilotParams.setup].keys():
+        pilotParams.commandExtensions = [str( pv ) for pv in pilotCommandsFileContent[pilotParams.setup]['Extensions']]
+    elif grid in pilotCommandsFileContent['Defaults']['Commands'].keys():
+      pilotParams.commands = [str( pv ) for pv in pilotCommandsFileContent['Defaults']['Commands'][grid]]
+    else:
+      pilotParams.commands = [str( pv ) for pv in pilotCommandsFileContent['Defaults']['Commands']['defaultList']]
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/dirac-pilot.py]: I would encapsulate the above logic in a separate function. And a documentation on what could be the format of the file is needed.

File: WorkloadManagementSystem/PilotAgent/pilotTools.py [hunk 4]
```diff
@@ -74,9 +74,10 @@ def retrieveUrlTimeout( url, fileName, log, timeout = 0 ):
       expectedBytes = 0
     data = remoteFD.read()
     if fileName:
-      localFD = open( fileName + '-local', "wb" )
-      localFD.write( data )
-      localFD.close()
+      if not os.path.exists( fileName + '-local' ):
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/pilotTools.py]: 'not' ?

**Comment by fstagni** [comment]
Needs unit test

File: WorkloadManagementSystem/PilotAgent/pilotCommands.py [hunk 5]
```diff
@@ -54,12 +54,16 @@ def execute(self):
...
                                    timeout = 120 )
       if not result:
-        self.log.error( "Failed to get pilot version, exiting ...")
+        self.log.error( "Failed to get pilot version, exiting ..." )
         sys.exit( 1 )
       fp = open( self.pp.pilotCFGFile + '-local', 'r' )
       pilotCFGFileContent = json.load( fp )
       fp.close()
-      pilotVersions = [str( pv ) for pv in pilotCFGFileContent[self.pp.setup]['Version']]
+      if self.pp.setup in pilotCFGFileContent.keys() and 'Version' in pilotCFGFileContent[self.pp.setup].keys():
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/pilotCommands.py]: You probably don't need .keys() here

File: WorkloadManagementSystem/PilotAgent/pilotCommands.py [hunk 6]
```diff
@@ -54,12 +54,16 @@ def execute(self):
...
+        pilotVersions = [str( pv ) for pv in pilotCFGFileContent[self.pp.setup]['Version']]
+      else:
+        pilotVersions = [str( pv ) for pv in pilotCFGFileContent['Defaults']['Version']]
+
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/pilotCommands.py]: Maybe put a comment explaining what's the logic here

**Comment by fstagni** [comment]
This PR should be moved to the Pilot repository. At the same time we should first of all understand if the solution with the JSON file is the best one (probably not).

**Comment by CinziaLu** [comment]
Moved.

**Pull request closed without merging**